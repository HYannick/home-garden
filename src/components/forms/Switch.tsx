/** @jsx jsx */
import React, {Fragment} from 'react';
import styled from "@emotion/styled";
import {jsx} from '@emotion/core';
import {Label} from "./Label";

interface SwitchProps {
  field?: any,
  form?: any,
  label?: string,
  onChange: (checked: boolean) => void
}

const SwitchWrapper = styled('label')`
  position: relative;
  display: inline-block;
  width: 7rem;
  height: 3.4rem;
`;

const HiddenInput = styled('input')`
  width: 0.01rem;
  height: 0.01rem;
  opacity: 0;
  overflow: hidden; 
  position: absolute;
  z-index: -1;
  &:checked + .slider {
    background-color: ${({theme}) => theme.palette.light};
    box-shadow: 0 0 0 0.1rem ${({theme}) => theme.palette.grey.darkest};
  }
  
  &:focus + .slider {
    box-shadow: 0 0 0 0.1rem ${({theme}) => theme.palette.grey.darkest};
  }
  
  &:checked + .slider:before {
    background-color: ${({theme}) => theme.palette.primary.light};
    transform: translateX(3.6rem);
  }
`;

const Slider = styled('span')`
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${({theme}) => theme.palette.light};
  box-shadow: 0 0 0 0.1rem ${({theme}) => theme.palette.grey.darkest};
  transition: .3s ease;
  border-radius: 4rem;
  
  &:before {
    position: absolute;
    content: "";
    height: 2.6rem;
    width: 2.6rem;
    left: 0.4rem;
    bottom: 0.4rem;
    background-color: ${({theme}) => theme.palette.danger.light};
    box-shadow: 0 0 0 0.1rem ${({theme}) => theme.palette.grey.darkest};
    transition: .3s ease;
    border-radius: 50%;
  }
`;

const Switch: React.FC<SwitchProps> = ({field, label, onChange}) => (
  <Fragment>
    <Label htmlFor={field.name}>{label}</Label>
    <SwitchWrapper htmlFor={field.name}>
      <input type="checkbox"/>
      <HiddenInput id={field.name} name={field.name} type="checkbox" onChange={(e) => onChange(e.target.checked)}/>
      <Slider className="slider round"/>
    </SwitchWrapper>
  </Fragment>
);

export default Switch;
