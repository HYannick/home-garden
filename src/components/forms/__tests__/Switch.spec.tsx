import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup, fireEvent } from '@testing-library/react';
import { renderWithTheme } from '../../../theme-wrapper';
import Switch from '../Switch';

afterEach(cleanup);
describe('Switch', () => {
  const props = {
    field: {
      name: 'puke'
    },
    onChange: jest.fn(),
    label: 'Should I puke?'
  };
  it('should render properly', () => {
    const { container } = renderWithTheme(
      <Switch {...props}/>
    );
    expect(container).toMatchSnapshot();
  });
  it('should render the right label', () => {
    const { getByText } = renderWithTheme(
      <Switch {...props}/>
    );
    expect(getByText(/^Should I puke/)).toHaveTextContent('Should I puke');
  });
  it('should trigger onChange', () => {
    const { getByText } = renderWithTheme(
      <Switch {...props}/>
    );

    fireEvent.click(getByText('Should I puke?'));
    expect(props.onChange).toHaveBeenCalledTimes(1);
  });
});
