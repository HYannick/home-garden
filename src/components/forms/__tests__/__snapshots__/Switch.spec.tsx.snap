// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render properly 1`] = `
<div>
  <label
    class="css-npa6bc"
    for="puke"
  >
    Should I puke?
  </label>
  <label
    class="css-qwug7p"
    for="puke"
  >
    <input
      type="checkbox"
    />
    <input
      checked=""
      class="css-x6ytko"
      id="puke"
      name="puke"
      type="checkbox"
    />
    <span
      class="slider round css-843eyc"
    />
  </label>
</div>
`;
